src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
-         SafepointSynchronize::is_at_safepoint(),
-         "must be executed at a safepoint");
-  if (JvmtiEnv::environments_might_exist()) {
-    JvmtiEnvIterator it;
-    for (JvmtiEnvBase* env = it.first(); env != NULL; env = it.next(env)) {
-      JvmtiTagMap* tag_map = env->tag_map_acquire();
-      if (tag_map != NULL && !tag_map->is_empty()) {
-        tag_map->do_weak_oops(is_alive, f);
-      }
+  // Hopefully quick exit.
+  if (!JvmtiEnv::environments_might_exist()) { return; }
+
+  JvmtiEnvIterator it;
+  for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+    JvmtiTagMap* tag_map = env->tag_map_acquire();
+    if (tag_map != NULL) {
+      tag_map->_needs_rehashing = true;
     }
   }
 }
 
-void JvmtiTagMap::do_weak_oops(BoolObjectClosure* is_alive, OopClosure* f) {
-
-  // does this environment have the OBJECT_FREE event enabled
-  bool post_object_free = env()->is_enabled(JVMTI_EVENT_OBJECT_FREE);
-
-  // counters used for trace message
-  int freed = 0;
-  int moved = 0;
-
-  JvmtiTagHashmap* hashmap = this->hashmap();
-
-  // reenable sizing (if disabled)
-  hashmap->set_resizing_enabled(true);
-
-  // if the hashmap is empty then we can skip it
-  if (hashmap->_entry_count == 0) {
-    return;
-  }
-
-  // now iterate through each entry in the table
+void JvmtiTagMap::gc_notification(size_t num_dead_entries) {
 
-  JvmtiTagHashmapEntry** table = hashmap->table();
-  int size = hashmap->size();
-
-  JvmtiTagHashmapEntry* delayed_add = NULL;
-
-  for (int pos = 0; pos < size; ++pos) {
-    JvmtiTagHashmapEntry* entry = table[pos];
-    JvmtiTagHashmapEntry* prev = NULL;
-
-    while (entry != NULL) {
-      JvmtiTagHashmapEntry* next = entry->next();
-
-      // has object been GC'ed
-      if (!is_alive->do_object_b(entry->object_raw())) {
-        // grab the tag
-        jlong tag = entry->tag();
-        guarantee(tag != 0, "checking");
-
-        // remove GC'ed entry from hashmap and return the
-        // entry to the free list
-        hashmap->remove(prev, pos, entry);
-        destroy_entry(entry);
-
-        // post the event to the profiler
-        if (post_object_free) {
-          JvmtiExport::post_object_free(env(), tag);
+  // No locks during VM bring-up (0 threads) and no safepoints after main
+  // thread creation and before VMThread creation (1 thread); initial GC
+  // verification can happen in that window which gets to here.
+  if (!JvmtiEnv::environments_might_exist()) { return; }
+
+  bool is_vm_thread = Thread::current()->is_VM_thread();
+  if (!is_vm_thread) {
+    if (num_dead_entries != 0) {
+      JvmtiEnvIterator it;
+      for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+        JvmtiTagMap* tag_map = env->tag_map_acquire();
+        if (tag_map != NULL) {
+          // Lock each hashmap from concurrent posting and cleaning
+          MutexLocker ml(tag_map->lock(), Mutex::_no_safepoint_check_flag);
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
```

Comment:
This could call unlink_and_post_locked instead of manually locking.

Reply:
Ok, 2 requests.  I can call that then and move the logging.